wandb/data_types.py

```diff
@@ -446,82 +578,191 @@ def add_row(self, *row):
-    pass
+        # The run under which this object is bound, if any.
+        self._run = None
+
+    @classmethod
+    def get_media_subdir(cls):
+        raise NotImplementedError
+
+    def is_bound(self):
+        return self._run is not None
+
+    def bind_to_run(self, run, key, step, id_=None):
+        """Bind this object to a particular Run.
+
+        Calling this function is necessary so that we have somewhere specific to
+        put the file associated with this object, from which other Runs can
+        refer to it.
+        """
+        if run is None:
+            raise TypeError('Argument "run" must not be None.')
+        if self.is_bound():
+            raise RuntimeError('Value is already bound to a Run: {}'.format(self))
+        self._run = run
+
+        # This is a flawed way of checking whether the file is already in
+        # the Run directory. It'd be better to check the actual directory
+        # components.
+        if not os.path.realpath(self._path).startswith(os.path.realpath(self._run.dir)):
+            base_path = os.path.join(self._run.dir, self.get_media_subdir())
+            util.mkdir_exists_ok(base_path)
+
+            if self._is_tmp:
+                if id_ is None:
+                    id_ = self._sha256[:8]
+
+                extension = os.path.splitext(self._path)[1]
+                new_path = os.path.join(base_path, '{}_{}_{}{}'.format(key, step, id_, extension))
+
+                shutil.move(self._path, new_path)
+
+                self._path = new_path
+                self._is_tmp = False
+            else:
+                basename, extension = os.path.splitext(os.path.basename(self._path))
+                new_path = os.path.join(base_path, '{}_{}{}'.format(basename, self._sha256[:8], extension))
+                shutil.copy(self._path, new_path)
+                self._path = new_path
 
+    def to_json(self):
+        """Get the JSON-friendly dict that represents this object.
 
-class Audio(IterableMedia):
-    MAX_AUDIO_COUNT = 100
+        Only works if `self.bind_to_run()` has previously been called.
 
-    def __init__(self, data, sample_rate=None, caption=None):
+        The resulting dict lets you load this object into other W&B runs.
         """
-        Accepts numpy array of audio data. 
+        if not self.is_bound():
+            raise RuntimeError('Value of type {} must be bound to a run with bind_to_run() before being serialized to JSON.'.format(type(self).__name__))
+
+        return {
+            '_type': 'file',  # TODO(adrian): This isn't (yet) a real media type we support on the frontend.
+            'path': os.path.relpath(self._path, self._run.dir),  # TODO(adrian): Convert this to a path with forward slashes.
+            'sha256': self._sha256,
+            'size': self._size,
+            'entity': self._run.entity,
+            'project': self._run.project_name(),
+            'run': self._run.name,
```

Comment:
Where is this actually getting persisted in our database?  Is it a new api call?

Reply:
It gets stored as either a JSON object or a string, depending on where you put it. In a Summary or History it gets stored as nested JSON (just like always). BigQuery doesn't have designated JSON columns, so it just gets stored as a string in there.

Another possibility in BigQuery (not sure about Redshift, etc.) is storing them in nested columns, because it lets you have subcolumns and multiple values in repeated columns. The advantage of doing it that way is that we'd be able to query it more naturally with BigQuery's SQL-ish syntax. The disadvantage is that we'd have to write code to do migrations on the table in order to add new fields or change their types. BigQuery does automatic migrations when you add normal fields, but it doesn't know how to handle nested fields. For now, at least, I'd prefer to keep it more flexible so we can change the design more easily.